<h3> Reading csv file using pandas</h3>
Pandas provides helper functions to read data from various file formats like CSV, Excel spreadsheets, HTML tables, JSON, SQL, and more.

In [1]:
# download italy-covid-daywise.txt

from urllib.request import urlretrieve

In [2]:
italy_covid_url = italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url, 'italy-covid-daywise.csv')

('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x7f97dc5c7ac0>)

to read file use read_csv method from pandas

In [3]:
!pip install pandas --upgrade --quiet

In [4]:
import pandas as pd 

In [5]:
covid_df = pd.read_csv('italy-covid-daywise.csv')

Data from the file is read and stored in a `DataFrame` object - one of the core data structures in Pandas for storing and working with tabular data. We typically use the `_df` suffix in the variable names for dataframes.

In [6]:
type(covid_df)

pandas.core.frame.DataFrame

In [7]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

Here's what we can tell by looking at the dataframe:

- The file provides four day-wise counts for COVID-19 in Italy
- The metrics reported are new cases, deaths, and tests
- Data is provided for 248 days: from Dec 12, 2019, to Sep 3, 2020

We can view some basic information about the data frame using the `.info` method.

In [8]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [9]:
# use .describe method to get more information of data
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

In [10]:
covid_df.columns # gives list of columns within the dataframe

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [11]:
covid_df.shape # gives dimensions 

(248, 4)

Here's a summary of the functions & methods we've looked at so far:

* `pd.read_csv` - Read data from a CSV file into a Pandas `DataFrame` object
* `.info()` - View basic infomation about rows, columns & data types
* `.describe()` - View statistical information about numeric columns
* `.columns` - Get the list of column names
* `.shape` - Get the number of rows & columns as a tuple

## Retrieving data from a data frame

The first thing you might want to do is retrieve data from this data frame, e.g., the counts of a specific day or the list of values in a particular column. To do this, it might help to understand the internal representation of data in a data frame. Conceptually, you can think of a dataframe as a dictionary of lists: keys are column names, and values are lists/arrays containing data for the respective columns. 

In [12]:
# pandas format is similar to this 
covid_data_dict = {
    'date':       ['2020-08-30', '2020-08-31', '2020-09-01', '2020-09-02', '2020-09-03'],
    'new_cases':  [1444, 1365, 996, 975, 1326],
    'new_deaths': [1, 4, 6, 8, 6],
    'new_tests': [53541, 42583, 54395, None, None]
}

Representing data in the above format has a few benefits:

* All values in a column typically have the same type of value, so it's more efficient to store them in a single array.
* Retrieving the values for a particular row simply requires extracting the elements at a given index from each column array.
* The representation is more compact (column names are recorded only once) compared to other formats that use a dictionary for each row of data (see the example below).

In [13]:
# Pandas format is not similar to this
covid_data_list = [
    {'date': '2020-08-30', 'new_cases': 1444, 'new_deaths': 1, 'new_tests': 53541},
    {'date': '2020-08-31', 'new_cases': 1365, 'new_deaths': 4, 'new_tests': 42583},
    {'date': '2020-09-01', 'new_cases': 996, 'new_deaths': 6, 'new_tests': 54395},
    {'date': '2020-09-02', 'new_cases': 975, 'new_deaths': 8 },
    {'date': '2020-09-03', 'new_cases': 1326, 'new_deaths': 6},
]

with dictionary of list analogy in mind you can now guess how to retrieve data from data frame. for example we can get list of values from specific column using the [] indexing notation.

In [14]:
covid_data_dict['new_cases']

[1444, 1365, 996, 975, 1326]

In [15]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

each column is represented using data structure called Series, which is essentially a numpy array with some extra method and properties

In [16]:
type(covid_df['new_cases'])

pandas.core.series.Series

In [17]:
# like arrays you can retrieve specific values with series using the indexing notation [].

covid_df['new_cases'][246]

975.0

Pandas also provide .at method to retrieve the element at a specific row and column directly 

In [18]:
covid_df.at[246,'new_cases']

975.0

Pandas also allows accessing column as properties of the dataframe using the '.' notation. However method only works for columns whoose names do not contain space or speciel characters

In [19]:
covid_df.new_cases

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

passing list of coulmns with in the indexing notation [] to access a subset of the data frame with just the given columns.

In [20]:
cases_df = covid_df[['date', 'new_cases']]
cases_df

date  new_cases
0    2019-12-31        0.0
1    2020-01-01        0.0
2    2020-01-02        0.0
3    2020-01-03        0.0
4    2020-01-04        0.0
..          ...        ...
243  2020-08-30     1444.0
244  2020-08-31     1365.0
245  2020-09-01      996.0
246  2020-09-02      975.0
247  2020-09-03     1326.0

[248 rows x 2 columns]

The new data frame cases_df is simply a "view" of the original data frame covid_df. Both point to the same data in the computer's memory. Changing any values inside one of them will also change the respective values in the other.

In [21]:
# copy() method for the  full copy of the dataframe

covid_df_copy = covid_df.copy()

data inside covid_df_copy is comletely seperate from the covid_df and changing values inside one of them will not affect the other

In [22]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

In [23]:
covid_df.loc[243]

date          2020-08-30
new_cases         1444.0
new_deaths           1.0
new_tests        53541.0
Name: 243, dtype: object

Each retrieved row is also a serires object

In [24]:
type(covid_df.loc[243])

pandas.core.series.Series

We can use `.head()` and `.tail()` method to view first and las rows of the data 


In [25]:
covid_df.head(5)

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN
3  2020-01-03        0.0         0.0        NaN
4  2020-01-04        0.0         0.0        NaN

In [26]:
covid_df.tail(5)

date  new_cases  new_deaths  new_tests
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

In [27]:
covid_df.at[0, 'new_tests']

nan

In [28]:
type(covid_df.at[0, 'new_tests'])

numpy.float64

In [29]:
# passing range to loc
covid_df.loc[108:113]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0

We can use `.sample()` method to retrieve random sample of rows from data frame

In [30]:
covid_df.sample(10)

date  new_cases  new_deaths  new_tests
167  2020-06-15      338.0        44.0    17463.0
28   2020-01-28        0.0         0.0        NaN
106  2020-04-15     2972.0       604.0        NaN
52   2020-02-21        0.0         0.0        NaN
59   2020-02-28      250.0         5.0        NaN
200  2020-07-18      231.0        11.0    27569.0
32   2020-02-01        0.0         0.0        NaN
145  2020-05-24      669.0       119.0    34206.0
50   2020-02-19        0.0         0.0        NaN
233  2020-08-20      642.0         7.0    49662.0

We have taken random sample but each rows original index is preserved

Here's a summary of the functions & methods we looked at in this section:

- `covid_df['new_cases']` - Retrieving columns as a `Series` using the column name
- `new_cases[243]` - Retrieving values from a `Series` using an index
- `covid_df.at[243, 'new_cases']` - Retrieving a single value from a data frame
- `covid_df.copy()` - Creating a deep copy of a data frame
- `covid_df.loc[243]` - Retrieving a row or range of rows of data from the data frame
- `head`, `tail`, and `sample` - Retrieving multiple rows of data from the data frame
- `covid_df.new_tests.first_valid_index` - Finding the first non-empty index in a series


## Analyzing data from data frames

Let's try to answer some questions about our data.

**Q: What are the total number of reported cases and deaths related to Covid-19 in Italy?**

Similar to Numpy arrays, a Pandas series supports the `sum` method to answer these questions.

In [31]:
total_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()

In [32]:
print('The number of reported cases is {} and total number of deaths is {}'. format(int(total_cases), int(total_deaths)))

The number of reported cases is 271515 and total number of deaths is 35497


**Q: What is the overall death rate (ratio of reported deaths to reported cases)**

In [33]:
death_rate = covid_df.new_deaths.sum()/covid_df.new_cases.sum()

In [34]:
print('overall reported death rate in italy is {:.2f}%'.format(death_rate*100))

overall reported death rate in italy is 13.07%


**Q: What is the overall number of tests conducted? A total of 935310 tests were conducted before daily test numbers were reported.**

In [35]:
initial_tests = 935310
total_tests = initial_tests + covid_df.new_tests.sum()

In [36]:
total_tests

5214766.0

**Q: What fraction of tests returned a positive result?**

In [37]:
positive_rate = total_cases / total_tests

In [38]:
print('{:.2f}% of tests led to positive diagnosis.' .format(positive_rate*100))

5.21% of tests led to positive diagnosis.


## Querying and sorting rows

Let's say we want only want to look at the days which had more than 1000 reported cases. We can use a boolean expression to check which rows satisfy this criterion.

In [39]:
high_new_cases = covid_df.new_cases > 1000

In [40]:
high_new_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

The boolean expression returns a series containing `True` and `False` booolean values you can use this series to select subset of rows from the original dataframe, corresponfing to the `True` values in the series

In [41]:
covid_df[high_new_cases]

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

In [42]:
high_cases_df = covid_df[covid_df.new_cases > 1000]

In [43]:
high_cases_df

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

only first and last 5 rows are displayed we can change some of the display options to view all the data

In [44]:
from IPython.display import display

with pd.option_context('display.max_rows', 100):
    display(covid_df[covid_df.new_cases > 1000])

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
74   2020-03-14     2547.0       252.0        NaN
75   2020-03-15     3497.0       173.0        NaN
76   2020-03-16     2823.0       370.0        NaN
77   2020-03-17     4000.0       347.0        NaN
78   2020-03-18     3526.0       347.0        NaN
79   2020-03-19     4207.0       473.0        NaN
80   2020-03-20     5322.0       429.0        NaN
81   2020-03-21     5986.0       625.0        NaN
82   2020-03-22     6557.0       795.0        NaN
83   2020-03-23     5560.0       649.0        NaN
84   2020-03-24     4789.0       601.0        NaN
85   2020-03-25     5249.0       743.0        NaN
86   2020-03-26     5210.0       685.0        NaN
87   2020-03-27     6153.0       660.0        NaN
88   2020-03-28     5959.0       971.0        NaN
89   2020-03-29     5974.0       887.0        NaN
90   2020-03-30     5217.0       758.0        NaN
91   2020-03-31     4050.0       810.0        NaN
92   2020-04-01     4053.0       839.0        NaN
93   2020-04-02     4782.0       727.0        NaN
94   2020-04-03     4668.0       760.0        NaN
95   2020-04-04     4585.0       764.0        NaN
96   2020-04-05     4805.0       681.0        NaN
97   2020-04-06     4316.0       527.0        NaN
98   2020-04-07     3599.0       636.0        NaN
99   2020-04-08     3039.0       604.0        NaN
100  2020-04-09     3836.0       540.0        NaN
101  2020-04-10     4204.0       612.0        NaN
102  2020-04-11     3951.0       570.0        NaN
103  2020-04-12     4694.0       619.0        NaN
104  2020-04-13     4092.0       431.0        NaN
105  2020-04-14     3153.0       564.0        NaN
106  2020-04-15     2972.0       604.0        NaN
107  2020-04-16     2667.0       578.0        NaN
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
115  2020-04-24     2646.0       464.0    95273.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
119  2020-04-28     1739.0       333.0    37554.0
120  2020-04-29     2091.0       382.0    38589.0
121  2020-04-30     2086.0       323.0    41441.0
122  2020-05-01     1872.0       285.0    43732.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
126  2020-05-05     1221.0       195.0    32211.0
127  2020-05-06     1075.0       236.0    37771.0
128  2020-05-07     1444.0       369.0    13665.0
129  2020-05-08     1401.0       274.0    45428.0
130  2020-05-09     1327.0       243.0    36091.0
131  2020-05-10     1083.0       194.0    31384.0
134  2020-05-13     1402.0       172.0    37049.0
236  2020-08-23     1071.0         3.0    47463.0
237  2020-08-24     1209.0         7.0    33358.0
240  2020-08-27     1366.0        13.0    57640.0
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

let's try to determine the days when the ratio of cases reported to tests conducted is higher than the overall `positive_rate`.

In [45]:
positive_rate

0.05206657403227681

In [46]:
high_ratio_df = covid_df[covid_df.new_cases / covid_df.new_tests > positive_rate]

In [47]:
high_ratio_df

date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
120  2020-04-29     2091.0       382.0    38589.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
128  2020-05-07     1444.0       369.0    13665.0

In [48]:
covid_df.new_cases / covid_df.new_tests

0           NaN
1           NaN
2           NaN
3           NaN
4           NaN
         ...   
243    0.026970
244    0.032055
245    0.018311
246         NaN
247         NaN
Length: 248, dtype: float64

Adding new column to the data frame

In [49]:
covid_df['positive_rate'] = covid_df.new_cases / covid_df.new_tests

In [50]:
covid_df

date  new_cases  new_deaths  new_tests  positive_rate
0    2019-12-31        0.0         0.0        NaN            NaN
1    2020-01-01        0.0         0.0        NaN            NaN
2    2020-01-02        0.0         0.0        NaN            NaN
3    2020-01-03        0.0         0.0        NaN            NaN
4    2020-01-04        0.0         0.0        NaN            NaN
..          ...        ...         ...        ...            ...
243  2020-08-30     1444.0         1.0    53541.0       0.026970
244  2020-08-31     1365.0         4.0    42583.0       0.032055
245  2020-09-01      996.0         6.0    54395.0       0.018311
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

[248 rows x 5 columns]

However, keep in mind that sometimes it takes a few days to get the results for a test, so we can't compare the number of new cases with the number of tests conducted on the same day. Any inference based on this `positive_rate` column is likely to be incorrect. It's essential to watch out for such subtle relationships that are often not conveyed within the CSV file and require some external context. It's always a good idea to read through the documentation provided with the dataset or ask for more information.

For now, let's remove the `positive_rate` column using the `drop` method.

In [51]:
covid_df.drop(columns=['positive_rate'], inplace = True)

In [52]:
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

### Sorting rows using column values

The rows can also be sorted by a specific column using `.sort_values`. Let's sort to identify the days with the highest number of cases, then chain it with the `head` method to list just the first ten results.

In [53]:
covid_df.sort_values('new_cases', ascending=False).head(10)

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN
83  2020-03-23     5560.0       649.0        NaN
80  2020-03-20     5322.0       429.0        NaN
85  2020-03-25     5249.0       743.0        NaN
90  2020-03-30     5217.0       758.0        NaN
86  2020-03-26     5210.0       685.0        NaN

It looks like last two weeks of march had the highest number of daily cases, Lets compare this to the days where the highest number of deaths were recorded

In [54]:
covid_df.sort_values('new_deaths', ascending=False).head(10) 

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN
95  2020-04-04     4585.0       764.0        NaN
94  2020-04-03     4668.0       760.0        NaN
90  2020-03-30     5217.0       758.0        NaN
85  2020-03-25     5249.0       743.0        NaN
93  2020-04-02     4782.0       727.0        NaN

Appears that daily death hit a peak just about weak after the peak in daily cases

Lets also look at days with the least number of cases. We might expect to see the first few days of the year on this list.

In [55]:
covid_df.sort_values('new_cases').head(10)

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
16   2020-01-16        0.0         0.0        NaN
26   2020-01-26        0.0         0.0        NaN
17   2020-01-17        0.0         0.0        NaN
28   2020-01-28        0.0         0.0        NaN
29   2020-01-29        0.0         0.0        NaN
30   2020-01-30        0.0         0.0        NaN
27   2020-01-27        0.0         0.0        NaN
36   2020-02-05        0.0         0.0        NaN
33   2020-02-02        0.0         0.0        NaN

It seems like the count of new cases on Jun 20, 2020, was -148, a negative number! Not something we might have expected, but that's the nature of real-world data. It could be a data entry error, or the government may have issued a correction to account for miscounting in the past.

In [56]:
covid_df.loc[169:175]

date  new_cases  new_deaths  new_tests
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20     -148.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

For now, let's assume this was indeed a data entry error. We can use one of the following approaches for dealing with the missing or faulty value:
1. Replace it with `0`.
2. Replace it with the average of the entire column
3. Replace it with the average of the values on the previous & next date
4. Discard the row entirely

Which approach you pick requires some context about the data and the problem. In this case, since we are dealing with data ordered by date, we can go ahead with the third approach.

You can use the `.at` method to modify a specific value within the dataframe.

In [57]:
covid_df.at[172, 'new_cases'] = (covid_df.at[171, 'new_cases'] + covid_df.at[173, 'new_cases'])/2

Here's a summary of the functions & methods we looked at in this section:

- `covid_df.new_cases.sum()` - Computing the sum of values in a column or series
- `covid_df[covid_df.new_cases > 1000]` - Querying a subset of rows satisfying the chosen criteria using boolean expressions
- `df['pos_rate'] = df.new_cases/df.new_tests` - Adding new columns by combining data from existing columns
- `covid_df.drop('positive_rate')` - Removing one or more columns from the data frame
- `sort_values` - Sorting the rows of a data frame using column values
- `covid_df.at[172, 'new_cases'] = ...` - Replacing a value within the data frame

## Working with dates

While we've looked at overall numbers for the cases, tests, positive rate, etc., it would also be useful to study these numbers on a month-by-month basis. The `date` column might come in handy here, as Pandas provides many utilities for working with dates.

In [58]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

The data type of date is currently `object`, so Pandas does not know that this column is a date. We can convert it into a `datetime` column using the `pd.to_datetime` method.

In [59]:
covid_df['date'] = pd.to_datetime(covid_df.date)

In [60]:
covid_df['date']

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

You can see that now it has data type datetime64. We can now extraxt different parts of the data into separate columns using the DateTimeIndex class 

In [61]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [62]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5
..         ...        ...         ...        ...   ...    ...  ...      ...
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3

[248 rows x 8 columns]

Let's check the overall metrics for May. We can query the rows for May, choose a subset of columns, and use the `sum` method to aggregate each selected column's values.

In [63]:
# Query the rows for may
covid_df_may = covid_df[covid_df.month == 5]

# Extract the subset of the columns to be aggregated 

covid_df_may_metrics = covid_df_may[['new_cases', 'new_deaths', 'new_tests']]

# Get the column-wise sum
covid_may_totals = covid_df_may_metrics.sum()

In [64]:
covid_may_totals

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

In [65]:
type(covid_may_totals)

pandas.core.series.Series

We can also combine above operations into single statement

In [66]:
covid_df[covid_df.month == 5][['new_cases', 'new_deaths', 'new_tests']].sum()

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

Lets check if number of cases reported on sundays is higher than the average cases reported everey day. This time, we might want to aggregate columns using the `.mean` method.

In [67]:
# overall average 
covid_df.new_cases.mean()

1096.6149193548388

In [68]:
# Average for sundays
covid_df[covid_df.weekday == 6].new_cases.mean()

1247.2571428571428

It seems like more cases were reported on sundays compared to other days

## Grouping and aggregation

As a next step, we might want to summarize the day-wise data and create a new dataframe with month-wise data. We can use the `groupby` function to create a group for each month, select the columns we wish to aggregate, and aggregate them using the `sum` method. 

In [69]:
covid_month_df = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].sum()

In [70]:
covid_month_df

new_cases  new_deaths  new_tests
month                                  
1            3.0         0.0        0.0
2          885.0        21.0        0.0
3       100851.0     11570.0        0.0
4       101852.0     16091.0   419591.0
5        29073.0      5658.0  1078720.0
6         8217.5      1404.0   830354.0
7         6722.0       388.0   797692.0
8        21060.0       345.0  1098704.0
9         3297.0        20.0    54395.0
12           0.0         0.0        0.0

The result is a new data frame that uses unique values from the column passed to `groupby` as the index. Grouping and aggregation is a powerful method for progressively summarizing data into smaller data frames.

Instead of aggregating by sum, you can also aggregate by other measures like mean. Let's compute the average number of daily new cases, deaths, and tests for each month.

In [71]:
covid_month_mean_df = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].mean()

In [72]:
covid_month_mean_df

new_cases  new_deaths     new_tests
month                                       
1         0.096774    0.000000           NaN
2        30.517241    0.724138           NaN
3      3253.258065  373.225806           NaN
4      3395.066667  536.366667  38144.636364
5       937.838710  182.516129  34797.419355
6       273.916667   46.800000  27678.466667
7       216.838710   12.516129  25732.000000
8       679.354839   11.129032  35442.064516
9      1099.000000    6.666667  54395.000000
12        0.000000    0.000000           NaN

Apart from grouping, another form of aggregation is the running or cumulative sum of cases, tests, or death up to each row's date. We can use the `cumsum` method to compute the cumulative sum of a column as a new series. Let's add three new columns: `total_cases`, `total_deaths`, and `total_tests`.

In [73]:
covid_df['total_cases'] = covid_df.new_cases.cumsum()

In [74]:
covid_df['total_deaths'] = covid_df.new_deaths.cumsum()

In [75]:
covid_df['total_tests'] = covid_df.new_tests.cumsum() + initial_tests

We've also included the initial test count in `total_test` to account for tests conducted before daily reporting was started. 

In [76]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths  total_tests  
0            0.0           0.0          NaN  
1            0.0           0.0          NaN  
2            0.0           0.0          NaN  
3            0.0           0.0          NaN  
4            0.0           0.0          NaN  
..           ...           ...          ...  
243     267298.5       35473.0    5117788.0  
244     268663.5       35477.0    5160371.0  
245     269659.5       35483.0    5214766.0  
246     270634.5       35491.0          NaN  
247     271960.5       35497.0          NaN  

[248 rows x 11 columns]

## Merging data from multiple sources

To determine other metrics like test per million, cases per million, etc., we require some more information about the country, viz. its population. Let's download another file `locations.csv` that contains health-related information for many countries, including Italy.

In [77]:
urlretrieve('https://gist.githubusercontent.com/aakashns/8684589ef4f266116cdce023377fc9c8/raw/99ce3826b2a9d1e6d0bde7e9e559fc8b6e9ac88b/locations.csv', 
            'locations.csv')

('locations.csv', <http.client.HTTPMessage at 0x7f97a1d48dc0>)

In [78]:
locations_df = pd.read_csv('locations.csv')

In [79]:
locations_df

location continent    population  life_expectancy  \
0      Afghanistan      Asia  3.892834e+07            64.83   
1          Albania    Europe  2.877800e+06            78.57   
2          Algeria    Africa  4.385104e+07            76.88   
3          Andorra    Europe  7.726500e+04            83.73   
4           Angola    Africa  3.286627e+07            61.15   
..             ...       ...           ...              ...   
207          Yemen      Asia  2.982597e+07            66.12   
208         Zambia    Africa  1.838396e+07            63.89   
209       Zimbabwe    Africa  1.486293e+07            61.49   
210          World       NaN  7.794799e+09            72.58   
211  International       NaN           NaN              NaN   

     hospital_beds_per_thousand  gdp_per_capita  
0                         0.500        1803.987  
1                         2.890       11803.431  
2                         1.900       13913.839  
3                           NaN             NaN  
4                           NaN        5819.495  
..                          ...             ...  
207                       0.700        1479.147  
208                       2.000        3689.251  
209                       1.700        1899.775  
210                       2.705       15469.207  
211                         NaN             NaN  

[212 rows x 6 columns]

In [80]:
locations_df[locations_df.location == "Italy"]

location continent  population  life_expectancy  \
97    Italy    Europe  60461828.0            83.51   

    hospital_beds_per_thousand  gdp_per_capita  
97                        3.18       35220.084

In [81]:
covid_df['location'] = "Italy"

In [82]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths  total_tests location  
0            0.0           0.0          NaN    Italy  
1            0.0           0.0          NaN    Italy  
2            0.0           0.0          NaN    Italy  
3            0.0           0.0          NaN    Italy  
4            0.0           0.0          NaN    Italy  
..           ...           ...          ...      ...  
243     267298.5       35473.0    5117788.0    Italy  
244     268663.5       35477.0    5160371.0    Italy  
245     269659.5       35483.0    5214766.0    Italy  
246     270634.5       35491.0          NaN    Italy  
247     271960.5       35497.0          NaN    Italy  

[248 rows x 12 columns]

We can now add the columns from locations_df into covid_df using the `.merge` method

In [83]:
merged_df = covid_df.merge(locations_df, on="location")

In [84]:
merged_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths  total_tests location continent  population  \
0            0.0           0.0          NaN    Italy    Europe  60461828.0   
1            0.0           0.0          NaN    Italy    Europe  60461828.0   
2            0.0           0.0          NaN    Italy    Europe  60461828.0   
3            0.0           0.0          NaN    Italy    Europe  60461828.0   
4            0.0           0.0          NaN    Italy    Europe  60461828.0   
..           ...           ...          ...      ...       ...         ...   
243     267298.5       35473.0    5117788.0    Italy    Europe  60461828.0   
244     268663.5       35477.0    5160371.0    Italy    Europe  60461828.0   
245     269659.5       35483.0    5214766.0    Italy    Europe  60461828.0   
246     270634.5       35491.0          NaN    Italy    Europe  60461828.0   
247     271960.5       35497.0          NaN    Italy    Europe  60461828.0   

     life_expectancy  hospital_beds_per_thousand  gdp_per_capita  
0              83.51                        3.18       35220.084  
1              83.51                        3.18       35220.084  
2              83.51                        3.18       35220.084  
3              83.51                        3.18       35220.084  
4              83.51                        3.18       35220.084  
..               ...                         ...             ...  
243            83.51                        3.18       35220.084  
244            83.51                        3.18       35220.084  
245            83.51                        3.18       35220.084  
246            83.51                        3.18       35220.084  
247            83.51                        3.18       35220.084  

[248 rows x 17 columns]

In [85]:
merged_df['cases_per_million'] = merged_df.total_cases * 1e6 / merged_df.population

In [86]:
merged_df['deaths_per_million'] = merged_df.total_deaths * 1e6 / merged_df.population

In [87]:
merged_df['tests_per_million'] = merged_df.total_tests * 1e6 / merged_df.population

In [88]:
merged_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday  \
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1   
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2   
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3   
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4   
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5   
..         ...        ...         ...        ...   ...    ...  ...      ...   
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6   
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0   
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1   
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2   
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3   

     total_cases  total_deaths  total_tests location continent  population  \
0            0.0           0.0          NaN    Italy    Europe  60461828.0   
1            0.0           0.0          NaN    Italy    Europe  60461828.0   
2            0.0           0.0          NaN    Italy    Europe  60461828.0   
3            0.0           0.0          NaN    Italy    Europe  60461828.0   
4            0.0           0.0          NaN    Italy    Europe  60461828.0   
..           ...           ...          ...      ...       ...         ...   
243     267298.5       35473.0    5117788.0    Italy    Europe  60461828.0   
244     268663.5       35477.0    5160371.0    Italy    Europe  60461828.0   
245     269659.5       35483.0    5214766.0    Italy    Europe  60461828.0   
246     270634.5       35491.0          NaN    Italy    Europe  60461828.0   
247     271960.5       35497.0          NaN    Italy    Europe  60461828.0   

     life_expectancy  hospital_beds_per_thousand  gdp_per_capita  \
0              83.51                        3.18       35220.084   
1              83.51                        3.18       35220.084   
2              83.51                        3.18       35220.084   
3              83.51                        3.18       35220.084   
4              83.51                        3.18       35220.084   
..               ...                         ...             ...   
243            83.51                        3.18       35220.084   
244            83.51                        3.18       35220.084   
245            83.51                        3.18       35220.084   
246            83.51                        3.18       35220.084   
247            83.51                        3.18       35220.084   

     cases_per_million  deaths_per_million  tests_per_million  
0             0.000000            0.000000                NaN  
1             0.000000            0.000000                NaN  
2             0.000000            0.000000                NaN  
3             0.000000            0.000000                NaN  
4             0.000000            0.000000                NaN  
..                 ...                 ...                ...  
243        4420.946386          586.700753       84644.943252  
244        4443.522614          586.766910       85349.238862  
245        4459.995818          586.866146       86248.897403  
246        4476.121695          586.998461                NaN  
247        4498.052887          587.097697                NaN  

[248 rows x 20 columns]

<h3> Writing data back to files</h3>
Before writing to file let us create a data frame which only contains column we wish to record.

In [89]:
result_df = merged_df[['date',
                       'new_cases',
                       'new_deaths',
                       'total_deaths',
                       'new_tests',
                       'total_tests',
                       'cases_per_million',
                       'deaths_per_million',
                       'tests_per_million']]

In [90]:
result_df

date  new_cases  new_deaths  total_deaths  new_tests  total_tests  \
0   2019-12-31        0.0         0.0           0.0        NaN          NaN   
1   2020-01-01        0.0         0.0           0.0        NaN          NaN   
2   2020-01-02        0.0         0.0           0.0        NaN          NaN   
3   2020-01-03        0.0         0.0           0.0        NaN          NaN   
4   2020-01-04        0.0         0.0           0.0        NaN          NaN   
..         ...        ...         ...           ...        ...          ...   
243 2020-08-30     1444.0         1.0       35473.0    53541.0    5117788.0   
244 2020-08-31     1365.0         4.0       35477.0    42583.0    5160371.0   
245 2020-09-01      996.0         6.0       35483.0    54395.0    5214766.0   
246 2020-09-02      975.0         8.0       35491.0        NaN          NaN   
247 2020-09-03     1326.0         6.0       35497.0        NaN          NaN   

     cases_per_million  deaths_per_million  tests_per_million  
0             0.000000            0.000000                NaN  
1             0.000000            0.000000                NaN  
2             0.000000            0.000000                NaN  
3             0.000000            0.000000                NaN  
4             0.000000            0.000000                NaN  
..                 ...                 ...                ...  
243        4420.946386          586.700753       84644.943252  
244        4443.522614          586.766910       85349.238862  
245        4459.995818          586.866146       86248.897403  
246        4476.121695          586.998461                NaN  
247        4498.052887          587.097697                NaN  

[248 rows x 9 columns]

to write data from data frame into file we can use the `to_csv` function

In [91]:
result_df.to_csv('result_csv', index=None)

`to_csv` function also includes additional column for storing the index of the dataframe by default, we pass index = None to turn off this behaviour.